### PR TITLE
feat: add Sublist Search Algorithm

### PR DESCRIPTION
**Pull request by foo290**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes: 

<a href="https://gitpod.io/#https://github.com/TheAlgorithms/C-Plus-Plus/pull/1513"></a>

### CODE REVIEW

File: search/sublist_search.cpp [hunk 1]
```diff
@@ -0,0 +1,289 @@
+/**
+ * @file
+ * @brief Implementation of the [Sublist Search
+ * Algorithm](https://www.geeksforgeeks.org/sublist-search-search-a-linked-list-in-another-list)
+ * algorithm.
```
Comment (Panquesito7) [inline search/sublist_search.cpp:5]:
```suggestion
 * @brief Implementation of the [Sublist Search
 * Algorithm](https://www.geeksforgeeks.org/sublist-search-search-a-linked-list-in-another-list)
```

File: search/sublist_search.cpp [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+ * of the first list with the first element of the second list.
+ *          * If the two values don't match, it goes to the next element of the
+ * second list. It does this until the two values match.
+ *
+ * @author [Nitin Sharma](https://github.com/foo290)
+ */
+
+#include <cassert>   // For assert
+#include <iostream>  // for IO operations
+#include <vector>    // For passing data as array
```
Comment (Panquesito7) [inline search/sublist_search.cpp:28]:
```suggestion
#include <cassert>   /// for assert
#include <iostream>  /// for IO operations
#include <vector>    /// for std::vector
```

**Review by Panquesito7 — COMMENTED** [review]
👍 Good work!

File: search/sublist_search.cpp [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+ *
+ * @author [Nitin Sharma](https://github.com/foo290)
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <vector>    /// for std::vector
+
+/**
+ * A Node structure representing single link Node in a linked list.
```
Comment (Panquesito7) [inline search/sublist_search.cpp:30]:
```suggestion
 * @brief A Node structure representing a single link Node in a linked list.
```

File: search/sublist_search.cpp [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+struct Node {
+    int data;
+    Node *next;
+};
+
+/**
+ * @brief Main searching function
+ * @param sublist A linked list which is supposed to be searched in mainList.
+ * @param mainList A linked list in which sublist will be searched.
+ * @returns bool If the sublist is found, returns true, else false.
```
Comment (Panquesito7) [inline search/sublist_search.cpp:41]:
```suggestion
 * @returns true if the sublist is found
 * @returns false if the sublist is NOT found
```

File: search/sublist_search.cpp [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        mainList = mainList->next;  // set the main pointer to the next element
+                                    // of the main list and repeat the algo.
+    }
+
+    // If the main list is exhausted, means sublist does not found, return false
+    return false;
+}
+
+/**
+ * @brief A simple function to print the linked list.
```
Comment (Panquesito7) [inline search/sublist_search.cpp:87]:
```suggestion
 * @brief A simple function to print the linked list
```

File: search/sublist_search.cpp [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+        log("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+            "~");
+    }
+};
+
+/**
+ * @brief A test function that runs various checks
+ * @returns void
+ */
+void test() {
```
Comment (Panquesito7) [inline search/sublist_search.cpp:254]:
```suggestion
static void test() {
```

File: search/sublist_search.cpp [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        log("[PASS] : TEST CASE 3 PASS!");
+        log("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+            "~");
+    }
+};
+
+/**
+ * @brief A test function that runs various checks
+ * @returns void
+ */
```
Comment (Panquesito7) [inline search/sublist_search.cpp:253]:
```suggestion
/**
 * @brief Self-test implementations
 * @returns void
 */
```

File: search/sublist_search.cpp [hunk 8]
```diff
@@ -0,0 +1,288 @@
...
+ * second list. It does this until the two values match.
+ *
+ * @author [Nitin Sharma](https://github.com/foo290)
+ */
+
+#include <cassert>   /// for assert
+#include <iostream>  /// for IO operations
+#include <vector>    /// for std::vector
+
+/**
```
Comment (Panquesito7) [inline search/sublist_search.cpp:29]:
```suggestion
/**
 * @namespace search
 * @brief Searching algorithms
 */
namespace search {
/**
 * @namespace sublist_search
 * @brief Functions for the [Sublist Search](https://www.geeksforgeeks.org/sublist-search-search-a-linked-list-in-another-list) implementation
/**
```

File: search/sublist_search.cpp [hunk 9]
```diff
@@ -0,0 +1,288 @@
...
+        if (head == nullptr) {
+            head = node;
+            tail = node;
+        } else {
+            tail->next = node;
+            tail = tail->next;
+        }
+    }
+    return head;
+}
```
Comment (Panquesito7) [inline search/sublist_search.cpp:121]:
```suggestion
}
}  // namespace sublist_search
}  // namespace search
```

File: search/sublist_search.cpp [hunk 10]
```diff
@@ -0,0 +1,288 @@
...
+
+/**
+ * A class, encapsulating the necessary test cases.
```
Comment (Panquesito7) [inline search/sublist_search.cpp:124]:
```suggestion
 * @brief a class encapsulating the necessary test cases.
```

File: search/sublist_search.cpp [hunk 11]
```diff
@@ -0,0 +1,288 @@
...
+            "~");
+        log("This is test case 1 for sublist search Algorithm : ");
+        log("Description:");
+        log("   EDGE CASE : Only contains one element");
+
+        std::vector<int> sublistData = {6};
+        std::vector<int> mainlistData = {2, 5, 6, 7, 8};
+
+        Node *sublistLL = makeLinkedList(sublistData);
+        Node *mainlistLL = makeLinkedList(mainlistData);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:159]:
```suggestion
        search::sublist_search::Node *sublistLL = search::sublist_search::makeLinkedList(sublistData);
        search::sublist_search::Node *mainlistLL = search::sublist_search::makeLinkedList(mainlistData);
```

File: search/sublist_search.cpp [hunk 12]
```diff
@@ -0,0 +1,288 @@
...
+
+        bool exists = sublistSearch(sublistLL, mainlistLL);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:161]:
```suggestion
        bool exists = search::sublist_search::sublistSearch(sublistLL, mainlistLL);
```

File: search/sublist_search.cpp [hunk 13]
```diff
@@ -0,0 +1,288 @@
...
+
+        int temp = 0;
+        for (int i = 45; i < 65; i++) {
+            // Inserts 20 elements in sublist
+            sublistData[temp] = i + 1;
+            temp++;
+        }
+
+        Node *sublistLL = makeLinkedList(sublistData);
+        Node *mainlistLL = makeLinkedList(mainlistData);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:200]:
```suggestion
        search::sublist_search::Node *sublistLL = search::sublist_search::makeLinkedList(sublistData);
        search::sublist_search::Node *mainlistLL = search::sublist_search::makeLinkedList(mainlistData);
```

File: search/sublist_search.cpp [hunk 14]
```diff
@@ -0,0 +1,288 @@
...
+
+        bool exists = sublistSearch(sublistLL, mainlistLL);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:202]:
```suggestion
        bool exists = search::sublist_search::sublistSearch(sublistLL, mainlistLL);
```

File: search/sublist_search.cpp [hunk 15]
```diff
@@ -0,0 +1,288 @@
...
+            mainlistData.push_back(i + 1);
+        }
+
+        for (int i = 45; i < 65; i++) {
+            // Inserts 20 elements in sublist
+            sublistData.push_back(i + 1);
+        }
+
+        Node *sublistLL = makeLinkedList(sublistData);
+        Node *mainlistLL = makeLinkedList(mainlistData);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:236]:
```suggestion
        search::sublist_search::Node *sublistLL = search::sublist_search::makeLinkedList(sublistData);
        search::sublist_search::Node *mainlistLL = search::sublist_search::makeLinkedList(mainlistData);
```

File: search/sublist_search.cpp [hunk 16]
```diff
@@ -0,0 +1,288 @@
...
+
+        bool exists = sublistSearch(sublistLL, mainlistLL);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:238]:
```suggestion
        bool exists = search::sublist_search::sublistSearch(sublistLL, mainlistLL);
```

File: search/sublist_search.cpp [hunk 17]
```diff
@@ -0,0 +1,288 @@
...
+    return head;
+}
+
+/**
+ * A class, encapsulating the necessary test cases.
+ */
+class TestCases {
+ private:
+    template <typename T>
+    void log(T msg) {
```
Comment (Panquesito7) [inline search/sublist_search.cpp:129]: Could you please add documentation for the following functions?

File: search/sublist_search.cpp [hunk 18]
```diff
@@ -0,0 +1,288 @@
...
+}
+
+/**
+ * @brief Main function
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main(int argc, char *argv[]) {
+    test();  // Executing various test cases first.
```
Comment (Panquesito7) [inline search/sublist_search.cpp:266]:
```suggestion
    test();  // run self-test implementations
```

File: search/sublist_search.cpp [hunk 19]
```diff
@@ -0,0 +1,288 @@
...
+
+    std::vector<int> mainlistData = {2, 5, 6, 7, 8};
+    std::vector<int> sublistData = {6, 8};
+    Node *mainlistLL = makeLinkedList(mainlistData);
+    Node *sublistLL = makeLinkedList(sublistData);
+
+    bool exists = sublistSearch(sublistLL, mainlistLL);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:273]:
```suggestion
    search::sublist_search::Node *mainlistLL = search::sublist_search::makeLinkedList(mainlistData);
    search::sublist_search::Node *sublistLL = search::sublist_search::makeLinkedList(sublistData);

    bool exists = search::sublist_search::sublistSearch(sublistLL, mainlistLL);
```

**Review by Panquesito7 — COMMENTED** [review]
Almost there! 😄

File: search/sublist_search.cpp (hunk 17 above)
Comment (Panquesito7) [reply search/sublist_search.cpp:144]: Still missing documentation (same for the other test-cases functions below).

File: search/sublist_search.cpp [hunk 20]
```diff
@@ -0,0 +1,338 @@
...
+ * Search](https://www.geeksforgeeks.org/sublist-search-search-a-linked-list-in-another-list)
+ * implementation
+ */
+namespace sublist_search {
+/**
+ * @brief A Node structure representing a single link Node in a linked list
+ */
+struct Node {
+    uint32_t data = 0;
+    Node *next{};
```
Comment (Panquesito7) [inline search/sublist_search.cpp:46]: Add a one-line description of what these variables are for (see the example below).
```c++
int64_t data = 0;       ///< The value/key of the node
```

File: search/sublist_search.cpp [hunk 21]
```diff
@@ -0,0 +1,338 @@
...
+        log("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+            "~");
+    }
+
+    /**
+     * @brief A test case which contains main list of 50 elements and sublist
+     * of 20.
+     * @returns void
+     * */
+    void testCase_3() {
```
Comment (Panquesito7) [inline search/sublist_search.cpp:258]: I'm not sure if these tests should be `static` too. @ayaankhan98, @mishraabhinn?

File: search/sublist_search.cpp [hunk 22]
```diff
@@ -0,0 +1,338 @@
...
+            sublistData.push_back(i + 1);
+        }
+
+        search::sublist_search::Node *sublistLL =
+            search::sublist_search::makeLinkedList(sublistData);
+        search::sublist_search::Node *mainlistLL =
+            search::sublist_search::makeLinkedList(mainlistData);
+
+        bool exists =
+            search::sublist_search::sublistSearch(sublistLL, mainlistLL);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:286]: Add a one-line description of what the variable is for (see the example below). Check other parts of the code, too.
```c++
int64_t data = 0;       ///< The value/key of the node.
```

File: search/sublist_search.cpp (hunk 21 above)
Comment (mishraabhinn) [reply search/sublist_search.cpp:258]: In this particular case, there is no reason why it must be static (based on the code provided).

Comment (Panquesito7) [reply search/sublist_search.cpp:258]: In tests cases, we use `static` for a Doxygen bug. Although, the main test function is already `static`, so I'm not sure if to keep it as-is or make it static. Let's see if the documentation is fine after merging this PR. 🙂

File: search/sublist_search.cpp [hunk 23]
```diff
@@ -0,0 +1,339 @@
...
+ * @brief Functions for the [Sublist
+ * Search](https://www.geeksforgeeks.org/sublist-search-search-a-linked-list-in-another-list)
+ * implementation
+ */
+namespace sublist_search {
+/**
+ * @brief A Node structure representing a single link Node in a linked list
+ */
+struct Node {
+    uint32_t data = 0;  // The key/value of the node
```
Comment (Panquesito7) [inline search/sublist_search.cpp:45]:
```suggestion
    uint32_t data = 0;  ///< the key/value of the node
```

File: search/sublist_search.cpp [hunk 24]
```diff
@@ -0,0 +1,339 @@
...
+    Node *next{};       // Pointer to the next node
```
Comment (Panquesito7) [inline search/sublist_search.cpp:46]:
```suggestion
    Node *next{};       ///< pointer to the next node
```

File: search/sublist_search.cpp [hunk 25]
```diff
@@ -0,0 +1,339 @@
+/**
+ * @file
+ * @brief Implementation of the [Sublist Search
+ * Algorithm](https://www.geeksforgeeks.org/sublist-search-search-a-linked-list-in-another-list)
+ * @details
```
Comment (Panquesito7) [inline search/sublist_search.cpp:5]:
```suggestion
 * @details
 *
 * ### Algorithm
```

File: search/sublist_search.cpp [hunk 26]
```diff
@@ -0,0 +1,341 @@
...
+        log("Test Cases over!");
+        std::cout << std::endl;
+    }
+
+    /**
+     * @brief A test case contains edge case, Only contains one element.
+     * @returns void
+     * */
+    void testCase_1() {
+        const bool expectedOutput = true;
```
Comment (Panquesito7) [inline search/sublist_search.cpp:178]: Add a one-line description of what the variable is for (see the example below).
```c++
int64_t data = 0;       ///< The value/key of the node.
```

File: search/sublist_search.cpp [hunk 27]
```diff
@@ -0,0 +1,341 @@
...
+        delete (mainlistLL);
+    }
+
+    /**
+     * @brief A test case which contains main list of 100 elements and sublist
+     * of 20.
+     * @returns void
+     * */
+    void testCase_2() {
+        const bool expectedOutput = true;
```
Comment (Panquesito7) [inline search/sublist_search.cpp:215]: Same here.

File: search/sublist_search.cpp [hunk 28]
```diff
@@ -0,0 +1,341 @@
...
+            "~");
+    }
+
+    /**
+     * @brief A test case which contains main list of 50 elements and sublist
+     * of 20.
+     * @returns void
+     * */
+    void testCase_3() {
+        const bool expectedOutput = false;
```
Comment (Panquesito7) [inline search/sublist_search.cpp:261]: Same here.

File: search/sublist_search.cpp [hunk 29]
```diff
@@ -0,0 +1,341 @@
...
+/**
+ * @brief Makes a dummy linked list for testing.
+ * @param data A vector of "int" containing the data that is supposed to be
+ * stored in nodes of linked list.
+ * @returns Node* A head pointer to the linked list.
+ */
+Node *makeLinkedList(const std::vector<uint64_t> &data) {
+    Node *head = nullptr;
+    Node *tail = nullptr;
+    for (int i : data) {
```
Comment (Panquesito7) [inline search/sublist_search.cpp:73]: Maybe add a bit of documentation here about what does this code does?

File: search/sublist_search.cpp [hunk 30]
```diff
@@ -0,0 +1,341 @@
...
+bool sublistSearch(Node *sublist, Node *mainList) {
+    if (sublist == nullptr || mainList == nullptr) {
+        return false;
+    }
+
+    Node *target_ptr = sublist;
+
+    while (mainList != nullptr) {
+        Node *main_ptr = mainList;  // Initialize main pointer to the current
+        // node of main list.
```
Comment (Panquesito7) [inline search/sublist_search.cpp:104]:
```suggestion
        Node *main_ptr = mainList;  /// Initialize main pointer to the current
                                    /// node of main list.
```

File: search/sublist_search.cpp [hunk 31]
```diff
@@ -0,0 +1,341 @@
...
+        const bool expectedOutput = true;
+
+        log("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+            "~");
+        log("This is test case 1 for sublist search Algorithm : ");
+        log("Description:");
+        log("   EDGE CASE : Only contains one element");
+
+        std::vector<uint64_t> sublistData = {6};
+        std::vector<uint64_t> mainlistData = {2, 5, 6, 7, 8};
```
Comment (Panquesito7) [inline search/sublist_search.cpp:187]: Same here.

File: search/sublist_search.cpp [hunk 32]
```diff
@@ -0,0 +1,341 @@
...
+        const bool expectedOutput = true;
+
+        log("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+            "~");
+        log("This is test case 2 for sublist search Algorithm : ");
+        log("Description:");
+        log("   contains main list of 100 elements and sublist of 20");
+
+        std::vector<uint64_t> sublistData(20);
+        std::vector<uint64_t> mainlistData(100);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:224]: Same here.

File: search/sublist_search.cpp [hunk 33]
```diff
@@ -0,0 +1,341 @@
...
+        const bool expectedOutput = false;
+
+        log("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~"
+            "~");
+        log("This is test case 3 for sublist search Algorithm : ");
+        log("Description:");
+        log("   contains main list of 50 elements and sublist of 20");
+
+        std::vector<uint64_t> sublistData(20);
+        std::vector<uint64_t> mainlistData(50);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:270]: Same here.

File: search/sublist_search.cpp [hunk 34]
```diff
@@ -0,0 +1,341 @@
...
+ */
+int main(int argc, char *argv[]) {
+    test();  // run self-test implementations
+
+    std::vector<uint64_t> mainlistData = {2, 5, 6, 7, 8};
+    std::vector<uint64_t> sublistData = {6, 8};
+    search::sublist_search::Node *mainlistLL =
+        search::sublist_search::makeLinkedList(mainlistData);
+    search::sublist_search::Node *sublistLL =
+        search::sublist_search::makeLinkedList(sublistData);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:323]: Same here.

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
It's just to add documentation for some one-line variables, and you're done! 😄

File: search/sublist_search.cpp [hunk 35]
```diff
@@ -0,0 +1,365 @@
+/**
+ * @file
+ * @brief Implementation of the [Sublist Search
+ * Algorithm](https://www.geeksforgeeks.org/sublist-search-search-a-linked-list-in-another-list)
+ * @details
+ *
+ * ### Algorithm
```
Comment (Panquesito7) [inline search/sublist_search.cpp:7]:
```suggestion
 * ### Algorithm
 *
```

File: search/sublist_search.cpp [hunk 36]
```diff
@@ -0,0 +1,365 @@
...
+/**
+ * @brief Main function
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main(int argc, char *argv[]) {
+    test();  /// run self-test implementations
+
+    std::vector<uint64_t> mainlistData = {2, 5, 6, 7, 8};
```
Comment (Panquesito7) [inline search/sublist_search.cpp:341]:
```suggestion
    std::vector<uint64_t> mainlistData = {2, 5, 6, 7, 8}; ///< Main list in which sublist is to be searched
```

File: search/sublist_search.cpp [hunk 37]
```diff
@@ -0,0 +1,365 @@
...
+    std::vector<uint64_t> sublistData = {6, 8};
```
Comment (Panquesito7) [inline search/sublist_search.cpp:342]:
```suggestion
    std::vector<uint64_t> sublistData = {6, 8}; ///< Sublist to be searched
```

File: search/sublist_search.cpp [hunk 38]
```diff
@@ -0,0 +1,365 @@
...
+
+    search::sublist_search::Node *mainlistLL =
+        search::sublist_search::makeLinkedList(mainlistData);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:345]:
```suggestion
        search::sublist_search::makeLinkedList(mainlistData); ///< Sublist to be searched
```

File: search/sublist_search.cpp [hunk 39]
```diff
@@ -0,0 +1,365 @@
...
+    search::sublist_search::Node *sublistLL =
+        search::sublist_search::makeLinkedList(sublistData);
```
Comment (Panquesito7) [inline search/sublist_search.cpp:347]:
```suggestion
        search::sublist_search::makeLinkedList(sublistData); ///< Main list in which sublist is to be
                                                            ///< searched
```

File: search/sublist_search.cpp [hunk 40]
```diff
@@ -0,0 +1,365 @@
...
+
+    bool exists = search::sublist_search::sublistSearch(
+        sublistLL, mainlistLL);  // boolean, if sublist exist or not
```
Comment (Panquesito7) [inline search/sublist_search.cpp:350]:
```suggestion
        sublistLL, mainlistLL);  ///< boolea to check if the sublist exists or not
```

File: search/sublist_search.cpp [hunk 41]
```diff
@@ -0,0 +1,365 @@
...
+
+    std::cout << "Sublist :" << std::endl;
```
Comment (Panquesito7) [inline search/sublist_search.cpp:352]:
```suggestion
    std::cout << "Sublist: " << std::endl;
```

File: search/sublist_search.cpp [hunk 42]
```diff
@@ -0,0 +1,365 @@
...
+    search::sublist_search::printLinkedList(sublistLL);
+
+    std::cout << "Main list : " << std::endl;
```
Comment (Panquesito7) [inline search/sublist_search.cpp:355]:
```suggestion
    std::cout << "Main list: " << std::endl;
```

**Review by Panquesito7 — CHANGES_REQUESTED** [review]
Almost there! 😄

File: search/sublist_search.cpp [hunk 43]
```diff
@@ -0,0 +1,367 @@
...
+}
+
+/**
+ * @brief Main function
+ * @param argc commandline argument count (ignored)
+ * @param argv commandline array of arguments (ignored)
+ * @returns 0 on exit
+ */
+int main(int argc, char *argv[]) {
+    test();  /// run self-test implementations
```
Comment (Panquesito7) [inline search/sublist_search.cpp:340]:
```suggestion
    test();  // run self-test implementations
```

**Review by Panquesito7 — APPROVED** [review]
Great work, @foo290! Thank you for your dedication and contributions to our community! 👍 🎉

**Review by mishraabhinn — APPROVED** [review]

**Review by ayaankhan98 — APPROVED** [review]
